builtins/web/fetch/headers.cpp

```diff
@@ -923,12 +1080,181 @@ unique_ptr<host_api::HttpHeaders> Headers::handle_clone(JSContext* cx, HandleObj
     return nullptr;
   }
 
-  auto handle = unique_ptr<host_api::HttpHeaders>(get_handle(self)->clone(guard(self)));
+  auto handle = unique_ptr<host_api::HttpHeaders>(get_handle(self)->clone());
   if (!handle) {
     api::throw_error(cx, FetchErrors::HeadersCloningFailed);
     return nullptr;
   }
   return handle;
 }
 
+BUILTIN_ITERATOR_METHODS(Headers)
+
+// Headers Iterator
+JSObject *HeadersIterator::create(JSContext *cx, HandleObject headers, uint8_t type) {
+  JSObject *self = JS_NewObjectWithGivenProto(cx, &class_, proto_obj);
+  if (!self) {
+    return nullptr;
+  }
+  SetReservedSlot(self, static_cast<uint32_t>(Slots::Type),
+                  JS::Int32Value(static_cast<int32_t>(type)));
+  SetReservedSlot(self, static_cast<uint32_t>(Slots::Cursor), JS::Int32Value(0));
+  SetReservedSlot(self, static_cast<uint32_t>(Slots::Headers), JS::ObjectValue(*headers));
+  return self;
+}
+
+const JSFunctionSpec HeadersIterator::static_methods[] = {
+    JS_FS_END,
+};
+
+const JSPropertySpec HeadersIterator::static_properties[] = {
+    JS_PS_END,
+};
+
+const JSFunctionSpec HeadersIterator::methods[] = {
+    JS_FN("next", HeadersIterator::next, 0, JSPROP_ENUMERATE),
+    JS_FS_END,
+};
+
+const JSPropertySpec HeadersIterator::properties[] = {
+    JS_PS_END,
+};
+
+bool HeadersIterator::init_class(JSContext *cx, JS::HandleObject global) {
+  JS::RootedObject iterator_proto(cx, JS::GetRealmIteratorPrototype(cx));
+  if (!iterator_proto)
+    return false;
+
+  if (!init_class_impl(cx, global, iterator_proto))
+    return false;
+
+  // Delete both the `HeadersIterator` global property and the
+  // `constructor` property on `HeadersIterator.prototype`. The latter
+  // because Iterators don't have their own constructor on the prototype.
+  return JS_DeleteProperty(cx, global, class_.name) &&
+         JS_DeleteProperty(cx, proto_obj, "constructor");
+}
+
+std::tuple<host_api::HostString, host_api::HostString> *
+Headers::get_index(JSContext *cx, JS::HandleObject self, size_t index) {
+  MOZ_ASSERT(is_instance(self));
+  std::vector<size_t> *headers_sort_list = static_cast<std::vector<size_t> *>(
```

Comment:
For both `HeadersList` and `HeadersSortList` it'd be very nice to introduce getter functions to eliminate the noise of retrieving them here and in a bunch of other places.

E.g. for `HeadersSortList`, I'd imagine that to be
```c++
Headers::headers_sort_list(JSObject* self) {
  MOZ_ASSERT(is_instance(self));
  auto *headers_sort_list = static_cast<std::vector<size_t> *>(
      JS::GetReservedSlot(self, static_cast<size_t>(Headers::Slots::HeadersSortList)).toPrivate());
  MOZ_ASSERT(headers_sort_list);
  return headers_sort_list;
}
```

reducing the first four lines of this function here to just
```c++
  auto *headers_sort_list = headers_sort_list(self);
```